oslabs-beta/DockerLocal
# Valid project names

## Description:
- Implemented project name validation and helper text for the creation of new projects. Added additional styling for the new project modal. Project names can contain alphanumeric characters, hypens, or underscores. Project names must be unique, contain less than 26 characters, and cannot be blank. If the user enters an invalid project name, the create project button is disabled, error messages & an error icon are displayed and the color of the UI elements in the create projects modal changes from blue to red.

## Changes I Made:
- [X] In  .gitignore, add  `/myProjects`
- [X] In App.sass, import Bulma components and helpers (typography, colors, form, message)
- [X] In ProjectSidebar.tsx, change projectNameValue state string to newProject state object which holds the project name string and isValid boolean to indicate whether user's current input into project name field is a valid project name. Add react components to display helper text message in new project modal. Implement conditional rendering for UI coloring, error message text, and error icon based on the value of newProject.isValid in state.
- [X] In projectHelper.ts, add checkValidName function. checkValidName is used to update the isValid boolean in the project sidebar state. It checks whether the project name is unique, of the right length, and includes only valid characters as detailed in Description above. 
- [X] In index.ts, add script to use Font Awesome. This is used for the error icon and will be used for other icons in the application.
- [X] In index.ts, use Prettier to format code. 


## How to Test:
- Run npm install. Run npm start. Use isLoggedIn button to navigate to the projects page. 

1. Click the Add Project button. The add project modal should open and the modal should look like Screenshot 1 (create project button is disabled but error messages and UI is blue). 

2. Enter a valid project name like 'demo' and the modal should look like Screenshot 2 (create project button is now enabled and UI is still blue). 

3. Enter a project name with invalid symbols or spaces like 'test %%@'. The modal should look like Screenshot 3 (error messages & error icon is displayed, UI is red). 

4. Enter a project name that already exists in the user's list of projects like 'DockerLocal'. The modal should look like Screenshot 4 (error messages & error icon is displayed, UI is red). 

5. Enter a project name that is longer than 25 characters or is blank. The modal should look like Screenshot 5 (error messages & error icon is displayed, UI is red). 


![onLoadProjectName](https://user-images.githubusercontent.com/7851187/87183624-5204ff00-c29b-11ea-9cdc-9d10cdd243e5.png)
**Screenshot 1**
![validProjectName](https://user-images.githubusercontent.com/7851187/87183744-8d9fc900-c29b-11ea-9205-769247718f37.png)
**Screenshot 2**
![errorProjectName](https://user-images.githubusercontent.com/7851187/87183775-9f816c00-c29b-11ea-8dba-a9446c739cc9.png)
**Screenshot 3**
![duplicateName](https://user-images.githubusercontent.com/7851187/87183784-a4deb680-c29b-11ea-8452-afa97570da2d.png)
**Screenshot 4**
![longProjectName](https://user-images.githubusercontent.com/7851187/87183791-a7411080-c29b-11ea-8379-c072d40b69f3.png)
**Screenshot 5**
